Index: /grass/trunk/display/d.grid/plot.c
===================================================================
--- /grass/trunk/display/d.grid/plot.c	(revision 68555)
+++ /grass/trunk/display/d.grid/plot.c	(revision 68556)
@@ -139,8 +139,19 @@
                     w = br - bl;
                     h = bt - bb;
+
+                    if (w > 0)
+                        w += 0.2 * fontsize * fabs(D_get_d_to_u_xconv());
+                    else  /* D_text() does not draw " ". */
+                        w = 0.8 * fontsize * fabs(D_get_d_to_u_xconv());
+                    if (h > 0)
+                        h += 0.2 * fontsize * fabs(D_get_d_to_u_yconv());
+                    else  /* D_text() does not draw " ". */
+                        h = 0.8 * fontsize * fabs(D_get_d_to_u_yconv());
+
                     bl = tx - w/2;
                     bt = ty + h/10;
                     br = tx + w + w/2;
                     bb = ty - h - h/10;
+
                     D_use_color(bgcolor);
                     D_box_abs(bl, bt, br, bb);
@@ -192,13 +203,11 @@
 
                     if (w > 0)
-                    w += 0.2 * fontsize;
-                    else
-                    /* D_text() does not draw " ". */
-                    w = 0.8 * fontsize;
+                        w += 0.2 * fontsize * fabs(D_get_d_to_u_xconv());
+                    else  /* D_text() does not draw " ". */
+                        w = 0.8 * fontsize * fabs(D_get_d_to_u_xconv());
                     if (h > 0)
-                    h += 0.2 * fontsize;
-                    else
-                    /* D_text() does not draw " ". */
-                    h = 0.8 * fontsize;
+                        h += 0.2 * fontsize * fabs(D_get_d_to_u_yconv());
+                    else  /* D_text() does not draw " ". */
+                        h = 0.8 * fontsize * fabs(D_get_d_to_u_yconv());
 
                     bl = tx - w/10;
